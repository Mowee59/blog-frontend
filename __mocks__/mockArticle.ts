import { Article_Preview } from "@/interfaces/article";
import { Media } from "@/interfaces/Media";

export const mockArticle: Article_Preview = {
	id: 2,
	attributes: {
		title: "Article test",
		slug: "article-test",
	
		publishedAt: "2024-09-10T16:16:35.724Z",
		description: "This is a test article summary",
		coverImage: {
			data: {
				id: 5,
				attributes: {
					name: "coverARticleTest.webp",
					alternativeText: "CoverImage",
					caption: "La description de l'image",
					width: 6000,
					height: 4000,
					formats: {
						thumbnail: {
							name: "thumbnail_coverARticleTest.webp",
							hash: "thumbnail_cover_A_Rticle_Test_44ad5169b1",
							ext: ".webp",
							mime: "image/webp",
							path: '',
							width: 234,
							height: 156,
							size: 4.9,
							url: "https://example.com/image.jpg"
						},
						small: {
							name: "small_coverARticleTest.webp",
							hash: "small_cover_A_Rticle_Test_44ad5169b1",
							ext: ".webp",
							mime: "image/webp",
							path: '',
							width: 500,
							height: 333,
							size: 13.1,
							url: "https://example.com/image1.jpg"
						},
						large: {
							name: "large_coverARticleTest.webp",
							hash: "large_cover_A_Rticle_Test_44ad5169b1",
							ext: ".webp",
							mime: "image/webp",
							path: '',
							width: 1000,
							height: 667,
							size: 33.6,
							url: "https://example.com/image2.jpg"
						},
						medium: {
							name: "medium_coverARticleTest.webp",
							hash: "medium_cover_A_Rticle_Test_44ad5169b1",
							ext: ".webp",
							mime: "image/webp",
							path: '',
							width: 750,
							height: 500,
							size: 22.59,
							url: "https://example.com/image3.jpg"
						}
					},
					hash: "cover_A_Rticle_Test_44ad5169b1",
					ext: ".webp",
					mime: "image/webp",
					size: 1282.64,
					url: "https://example.com/imageOriginal.jpg",
					previewUrl: '',
					provider: "local",
					createdAt: "2024-09-18T18:00:47.473Z",
					updatedAt: "2024-09-25T17:35:19.841Z"
				}
			}
		},
		tags: {
			data: [
				{
					id: 1,
					attributes: {
						cover: { data: {} as Media },
						name: "Algorithmie",
						createdAt: "2024-09-11T08:17:24.849Z",
						updatedAt: "2024-09-27T10:18:29.586Z",
						locale: "fr",
						description: "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Fusce mattis id augue non feugiat. Quisque rhoncus diam ornare ex feugiat, ac commodo odio aliquam. Vestibulum condimentum auctor tortor, ut laoreet lorem finibus et. Maecenas vel pulvinar eros, in porttitor metus. Donec tempor quis diam vitae interdum. Quisque mollis finibus purus, ac malesuada lectus auctor ut. In nisi tellus, ultricies non aliquam nec, sodales id turpis."
					}
				},
				{
					id: 3,
					attributes: {
						cover: { data: {} as Media },
						name: "Javascript",
						createdAt: "2024-09-19T13:58:51.645Z",
						updatedAt: "2024-09-27T10:18:36.955Z",
						locale: "fr",
						description: "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Fusce mattis id augue non feugiat. Quisque rhoncus diam ornare ex feugiat, ac commodo odio aliquam. Vestibulum condimentum auctor tortor, ut laoreet lorem finibus et. Maecenas vel pulvinar eros, in porttitor metus. Donec tempor quis diam vitae interdum. Quisque mollis finibus purus, ac malesuada lectus auctor ut. In nisi tellus, ultricies non aliquam nec, sodales id turpis."
					}
				},
				{
					id: 4,
					attributes: {
						cover: { data: {} as Media },
						name: "React",
						createdAt: "2024-09-20T10:30:18.922Z",
						updatedAt: "2024-09-27T10:18:40.751Z",
						locale: "fr",
						description: "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Fusce mattis id augue non feugiat. Quisque rhoncus diam ornare ex feugiat, ac commodo odio aliquam. Vestibulum condimentum auctor tortor, ut laoreet lorem finibus et. Maecenas vel pulvinar eros, in porttitor metus. Donec tempor quis diam vitae interdum. Quisque mollis finibus purus, ac malesuada lectus auctor ut. In nisi tellus, ultricies non aliquam nec, sodales id turpis."
					}
				}
			]
		},
	
	}}