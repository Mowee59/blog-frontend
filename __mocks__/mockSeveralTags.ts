// New file: __mocks__/mockSeveralTags.ts
import { Article, Article_NoRelations } from '@/interfaces/Article';
import { Tag } from '@/interfaces/Tag';

export const mockSeveralTags: Tag[] = [
  {
    id: 1,
    attributes: {
      name: "React",
      locale: "fr",
      createdAt: "2024-09-11T08:17:24.849Z",
      updatedAt: "2024-09-27T10:18:29.586Z",
      description: "Une librairie Javascript pour construire des interfaces utilisateur",
      cover: {
        data: {
          id: 3,
          attributes: {
            name: "react.webp",
            alternativeText: "Image de couverture catégorie React",
            caption: "React",
            width: 6000,
            height: 4000,
            formats: {
              thumbnail: {
                name: "thumbnail_react.webp",
                hash: "thumbnail_react_8111d1aa1b",
                ext: ".webp",
                mime: "image/webp",
                path: '',
                width: 234,
                height: 156,
                size: 3.86,
                url: "/uploads/thumbnail_react_8111d1aa1b.webp"
              },
              small: {
                name: "small_react.webp",
                hash: "small_react_8111d1aa1b",
                ext: ".webp",
                mime: "image/webp",
                path: '',
                width: 300,
                height: 200,
                size: 5.21,
                url: "/uploads/small_react_8111d1aa1b.webp"
              },
              medium: {
                name: "medium_react.webp",
                hash: "medium_react_8111d1aa1b",
                ext: ".webp",
                mime: "image/webp",
                path: '',
                width: 750,
                height: 500,
                size: 15.35,
                url: "/uploads/medium_react_8111d1aa1b.webp"
              },
              large: {
                name: "large_react.webp",
                hash: "large_react_8111d1aa1b",
                ext: ".webp",
                mime: "image/webp",
                path: '',
                width: 1000,
                height: 667,
                size: 21.93,
                url: "/uploads/large_react_8111d1aa1b.webp"
              }
            },
            hash: "react_8111d1aa1b",
            ext: ".webp",
            mime: "image/webp",
            size: 3633.27,
            url: "/uploads/react_8111d1aa1b.webp",
            previewUrl: '',
            provider: "local",
            createdAt: "2024-09-11T08:17:11.574Z",
            updatedAt: "2024-09-20T09:36:16.491Z"
          }
        }
      },
      articles: {
        data: [{} as Article, {} as Article, {} as Article]
      },
      localizations: {
        data: []
      }
    }
  },
  {
    id: 2,
    attributes: {
      name: "TypeScript",
      locale: "fr",
      createdAt: "2024-09-12T10:20:30.123Z",
      updatedAt: "2024-09-28T11:22:33.456Z",
      description: "Un sur-ensemble typé de JavaScript qui se compile en JavaScript pur",
      cover: {
        data: {
          id: 4,
          attributes: {
            name: "typescript.webp",
            alternativeText: "Image de couverture catégorie TypeScript",
            caption: "TypeScript",
            width: 6000,
            height: 4000,
            formats: {
              thumbnail: {
                name: "thumbnail_typescript.webp",
                hash: "thumbnail_typescript_9222e2bb2c",
                ext: ".webp",
                mime: "image/webp",
                path: '',
                width: 234,
                height: 156,
                size: 4.12,
                url: "/uploads/thumbnail_typescript_9222e2bb2c.webp"
              },
              small: {
                name: "small_typescript.webp",
                hash: "small_typescript_9222e2bb2c",
                ext: ".webp",
                mime: "image/webp",
                path: '',
                width: 300,
                height: 200,
                size: 5.67,
                url: "/uploads/small_typescript_9222e2bb2c.webp"
              },
              medium: {
                name: "medium_typescript.webp",
                hash: "medium_typescript_9222e2bb2c",
                ext: ".webp",
                mime: "image/webp",
                path: '',
                width: 750,
                height: 500,
                size: 16.89,
                url: "/uploads/medium_typescript_9222e2bb2c.webp"
              },
              large: {
                name: "large_typescript.webp",
                hash: "large_typescript_9222e2bb2c",
                ext: ".webp",
                mime: "image/webp",
                path: '',
                width: 1000,
                height: 667,
                size: 24.56,
                url: "/uploads/large_typescript_9222e2bb2c.webp"
              }
            },
            hash: "typescript_9222e2bb2c",
            ext: ".webp",
            mime: "image/webp",
            size: 3987.65,
            url: "/uploads/typescript_9222e2bb2c.webp",
            previewUrl: '',
            provider: "local",
            createdAt: "2024-09-12T10:20:15.789Z",
            updatedAt: "2024-09-21T12:34:56.789Z"
          }
        }
      },
      articles: {
        data: [{} as Article, {} as Article]
      },
      localizations: {
        data: []
      }
    }
  },
  {
    id: 3,
    attributes: {
      name: "Next.js",
      locale: "fr",
      createdAt: "2024-09-13T14:25:36.789Z",
      updatedAt: "2024-09-29T15:27:39.012Z",
      description: "Un framework React pour la production",
      cover: {
        data: {
          id: 5,
          attributes: {
            name: "nextjs.webp",
            alternativeText: "Image de couverture catégorie Next.js",
            caption: "Next.js",
            width: 6000,
            height: 4000,
            formats: {
              thumbnail: {
                name: "thumbnail_nextjs.webp",
                hash: "thumbnail_nextjs_3333f3cc3d",
                ext: ".webp",
                mime: "image/webp",
                path: '',
                width: 234,
                height: 156,
                size: 3.98,
                url: "/uploads/thumbnail_nextjs_3333f3cc3d.webp"
              },
              small: {
                name: "small_nextjs.webp",
                hash: "small_nextjs_3333f3cc3d",
                ext: ".webp",
                mime: "image/webp",
                path: '',
                width: 300,
                height: 200,
                size: 5.43,
                url: "/uploads/small_nextjs_3333f3cc3d.webp"
              },
              medium: {
                name: "medium_nextjs.webp",
                hash: "medium_nextjs_3333f3cc3d",
                ext: ".webp",
                mime: "image/webp",
                path: '',
                width: 750,
                height: 500,
                size: 15.87,
                url: "/uploads/medium_nextjs_3333f3cc3d.webp"
              },
              large: {
                name: "large_nextjs.webp",
                hash: "large_nextjs_3333f3cc3d",
                ext: ".webp",
                mime: "image/webp",
                path: '',
                width: 1000,
                height: 667,
                size: 22.76,
                url: "/uploads/large_nextjs_3333f3cc3d.webp"
              }
            },
            hash: "nextjs_3333f3cc3d",
            ext: ".webp",
            mime: "image/webp",
            size: 3845.92,
            url: "/uploads/nextjs_3333f3cc3d.webp",
            previewUrl: '',
            provider: "local",
            createdAt: "2024-09-13T14:25:30.123Z",
            updatedAt: "2024-09-22T16:38:59.012Z"
          }
        }
      },
      articles: {
        data: [{} as Article, {} as Article, {} as Article, {} as Article]
      },
      localizations: {
        data: []
      }
    }
  }
];