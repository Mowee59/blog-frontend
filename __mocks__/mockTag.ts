import { Article, Article_NoRelations } from '@/interfaces/Article';
import { Tag } from '@/interfaces/Tag';

export const mockTag: Tag = {
  id: 1,
  attributes: {
    name: "React",
    locale: "fr",
    createdAt: "2024-09-11T08:17:24.849Z",
    updatedAt: "2024-09-27T10:18:29.586Z",
    description: "Une librairie Javascript pour construire des interfaces utilisateur",
    cover: {
      data: {
        id: 3,
        attributes: {
          name: "algo.webp",
          alternativeText: "Image de couverture cat√©gorie Algorithmie",
          caption: "Alogrithmie",
          width: 6000,
          height: 4000,
          formats: {
            thumbnail: {
              name: "thumbnail_algo.webp",
              hash: "thumbnail_algo_8111d1aa1b",
              ext: ".webp",
              mime: "image/webp",
              path: '',
              width: 234,
              height: 156,
              size: 3.86,
              url: "/uploads/thumbnail_algo_8111d1aa1b.webp"
            },
            small: {
              name: "small_algo.webp",
              hash: "small_algo_8111d1aa1b",
              ext: ".webp",
              mime: "image/webp",
              path: '',
              width: 300,
              height: 200,
              size: 5.21,
              url: "/uploads/small_algo_8111d1aa1b.webp"
            },
            medium: {
              name: "medium_algo.webp",
              hash: "medium_algo_8111d1aa1b",
              ext: ".webp",
              mime: "image/webp",
              path: '',
              width: 750,
              height: 500,
              size: 15.35,
              url: "/uploads/medium_algo_8111d1aa1b.webp"
            },
            large: {
              name: "large_algo.webp",
              hash: "large_algo_8111d1aa1b",
              ext: ".webp",
              mime: "image/webp",
              path: '',
              width: 1000,
              height: 667,
              size: 21.93,
              url: "/uploads/large_algo_8111d1aa1b.webp"
            }
          },
          hash: "algo_8111d1aa1b",
          ext: ".webp",
          mime: "image/webp",
          size: 3633.27,
          url: "/uploads/algo_8111d1aa1b.webp",
          previewUrl: '',
          provider: "local",
          createdAt: "2024-09-11T08:17:11.574Z",
          updatedAt: "2024-09-20T09:36:16.491Z"
        }
      }
    },
    articles: {
      data: [  {} as Article, {} as Article , {} as Article]
    },
    localizations: {
      data: []
    }
  }
}

