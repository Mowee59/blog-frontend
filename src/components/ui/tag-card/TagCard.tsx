import { Tag } from "@/interfaces/tag";
import React from "react";
import Image from "next/image";
import { fetchArticlesByTagName } from "@/libs/api";

//Typing the component props
type TagCardProps = {
  tag: Tag;
};

const TagCard = async (props: TagCardProps) => {
  // Assigning the cover image attributes to a variable
  const coverImageAttributes = props.tag.attributes.cover.data.attributes;

  // Asssigning the thumbnail to a varialbe
  const coverImageThumbnail = coverImageAttributes.formats.thumbnail;

  // Fetching the articles related to the tag
  // We are adding  the 'fields[0]' query parameter to filter all list car we are only interested in metadata ( to get the number of articles related to the tag)
  const articles = await fetchArticlesByTagName(
    props.tag.attributes.name,
    "fields[0]",
  );

  return (
    <article className="flex h-[225px] w-full flex-col gap-5">
      <div className="relative h-[150px]">
        <Image
          src={`${process.env.STRAPI_URL}${coverImageThumbnail.url}`}
          alt={coverImageAttributes.alternativeText}
          fill
          placeholder="blur"
          blurDataURL={`${process.env.STRAPI_URL}${coverImageAttributes.previewUrl}`}
          objectFit="cover"
          className=" rounded-xl"
        ></Image>
      </div>
      <div className="flex flex-col gap-1">
        <h3 className="text-xl font-medium leading-snug text-neutral-900 dark:text-neutral-300">{`#${props.tag.attributes.name}`}</h3>
        <h4 className="text-sm font-medium leading-4 text-[#3d3d3d] dark:text-neutral-400">{`${articles.meta.pagination?.total} posts`}</h4>
      </div>
    </article>
  );
};

export default TagCard;
