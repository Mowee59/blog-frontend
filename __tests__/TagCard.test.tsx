import React from 'react';
import { render, screen } from '@testing-library/react';
import TagCard from '../src/components/ui/tag-card/TagCard';
import { mockTag } from '../__mocks__/mockTag';


describe('TagCard', () => {
  it('renders the tag name', () => {
    render(<TagCard tag={mockTag} />);
    expect(screen.getByText('#React')).toBeInTheDocument();
  });


  it('displays the correct number of articles', () => {
    render(<TagCard tag={mockTag} />);
    expect(screen.getByText('3 Articles')).toBeInTheDocument();
  });

  it('links to the correct tag page', () => {
    render(<TagCard tag={mockTag} />);
    const linkElement = screen.getByRole('link');
    expect(linkElement).toHaveAttribute('href', '/tags/react');
  });

  it('matches snapshot', () => {
    const { asFragment } = render(<TagCard tag={mockTag} />);
    expect(asFragment()).toMatchSnapshot();
  });
});