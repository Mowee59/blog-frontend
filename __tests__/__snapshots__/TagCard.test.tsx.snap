// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagCard matches snapshot 1`] = `
<DocumentFragment>
  <article
    class="flex h-[225px] w-full flex-col gap-5"
  >
    <a
      href="/tags/react"
    >
      <div>
        <div
          class="relative h-[150px]"
        >
          <img
            alt="Image de couverture catÃ©gorie Algorithmie"
            class=" rounded-xl object-cover"
            src="/uploads/thumbnail_algo_8111d1aa1b.webp"
          />
        </div>
        <div
          class="flex flex-col gap-1"
        >
          <h3
            class="text-xl font-medium leading-snug text-neutral-900 dark:text-neutral-300"
          >
            #React
          </h3>
          <h4
            class="text-sm font-medium leading-4 text-[#3d3d3d] dark:text-neutral-400"
          >
            3 Articles
          </h4>
        </div>
      </div>
    </a>
  </article>
</DocumentFragment>
`;
